File: nf-dream/scripts/lint.py
#!/usr/bin/env python3
"""nf-dream `lint` mode — read-only health check.

Folder-wide regardless of project filter (per SKILL.md: lint is the exception).
Reports issues, never writes.

Checks performed:
- MEMORY.md size > 20 KB
- Any file > 30 KB
- Files missing YAML frontmatter
- Files missing `name` / `description` (top-level)
- Files missing `metadata.type` / `metadata.project` / `metadata.cwd`
- Files with MISPLACED top-level keys (`type:` / `project:` / `cwd:` / `tags:` / `related:` at top
  instead of under `metadata:` — silently ignored by some tooling, see user-memory-frontmatter rule)
- MEMORY.md lines linking to files that don't exist
- Files that exist on disk but have no MEMORY.md entry
- archive/README.md entries pointing to files that no longer exist

Usage:
  lint.py <scope-folder> [--json]
"""
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path

TOP_RESERVED = {"name", "description", "created", "originSessionId", "last_read", "session_id"}
METADATA_KEYS = {"type", "project", "cwd", "tags", "related"}
SKIP_BASENAMES = {"MEMORY.md"}
HANDOFF_PATTERN = re.compile(r"^HANDOFF([-_].+)?\.md$", re.IGNORECASE)


def parse_frontmatter(text: str):
    if not text.startswith("---\n"):
        return None
    lines = text.split("\n")
    end = None
    for i in range(1, len(lines)):
        if lines[i] == "---":
            end = i
            break
    if end is None:
        return None
    return lines[1:end], "\n".join(lines[end + 1:])


def analyze_frontmatter(fm_lines: list[str]) -> dict:
    """Return dict of checks for one file's frontmatter."""
    has = {
        "name": False,
        "description": False,
        "metadata_block": False,
        "metadata.type": False,
        "metadata.project": False,
        "metadata.cwd": False,
    }
    misplaced_top: list[str] = []
    in_md = False
    md_indent_active = False
    i = 0
    while i < len(fm_lines):
        line = fm_lines[i]
        if line.strip() == "":
            i += 1
            continue
        m_top = re.match(r"^([A-Za-z_][A-Za-z0-9_]*):(.*)$", line)
        m_nested = re.match(r"^  ([A-Za-z_][A-Za-z0-9_]*):(.*)$", line)
        if m_top:
            key = m_top.group(1)
            md_indent_active = False
            if key == "name":
                has["name"] = True
            elif key == "description":
                has["description"] = True
            elif key == "metadata":
                has["metadata_block"] = True
                md_indent_active = True
            elif key in METADATA_KEYS:
                misplaced_top.append(key)
            i += 1
            continue
        if m_nested and (in_md or md_indent_active):
            nkey = m_nested.group(1)
            if nkey == "type":
                has["metadata.type"] = True
            elif nkey == "project":
                has["metadata.project"] = True
            elif nkey == "cwd":
                has["metadata.cwd"] = True
            i += 1
            continue
        if line.startswith("  ") and md_indent_active:
            i += 1
            continue
        md_indent_active = False
        i += 1
    return {"has": has, "misplaced_top": misplaced_top}


def main() -> int:
    parser = argparse.ArgumentParser(description="nf-dream lint")
    parser.add_argument("scope")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    scope = Path(args.scope).expanduser().resolve()
    if not scope.is_dir():
        print(f"ERROR: {scope} is not a directory", file=sys.stderr)
        return 1

    findings: list[dict] = []

    memory_md = scope / "MEMORY.md"
    if memory_md.exists():
        size_kb = memory_md.stat().st_size / 1024
        if size_kb > 20:
            findings.append({"kind": "memory_md_large", "size_kb": round(size_kb, 1)})

    files = sorted(scope.glob("*.md"))
    on_disk = {f.name for f in files if f.name not in SKIP_BASENAMES}

    indexed: set[str] = set()
    broken_links: list[str] = []
    if memory_md.exists():
        # Match Markdown links `[label](target.md)` only — not bare parenthesised text.
        for m in re.finditer(r"\[[^\]]+\]\(([^)]+\.md)\)", memory_md.read_text()):
            link = m.group(1)
            indexed.add(link)
            if not (scope / link).exists() and not link.startswith("archive/"):
                broken_links.append(link)

    # Files that exist on disk but have no MEMORY.md entry. Skip handoff files (they are deliverables
    # for fresh sessions, not memory entries — they reference MEMORY.md, not the other way around).
    candidates_for_index = {n for n in on_disk if not HANDOFF_PATTERN.match(n)}
    missing_index = candidates_for_index - indexed
    if broken_links:
        findings.append({"kind": "memory_md_broken_links", "links": broken_links})
    if missing_index:
        findings.append({"kind": "files_not_in_memory_md", "files": sorted(missing_index)})

    for f in files:
        if f.name in SKIP_BASENAMES:
            continue
        size_kb = f.stat().st_size / 1024
        if size_kb > 30:
            findings.append({"kind": "file_large", "file": f.name, "size_kb": round(size_kb, 1)})
        text = f.read_text()
        parsed = parse_frontmatter(text)
        if parsed is None:
            findings.append({"kind": "no_frontmatter", "file": f.name})
            continue
        fm_lines, _ = parsed
        info = analyze_frontmatter(fm_lines)
        missing = [k for k, v in info["has"].items() if not v]
        if missing:
            findings.append({"kind": "missing_fields", "file": f.name, "missing": missing})
        if info["misplaced_top"]:
            findings.append({"kind": "misplaced_top_level", "file": f.name, "keys": info["misplaced_top"]})

    archive_readme = scope / "archive" / "README.md"
    if archive_readme.exists():
        for m in re.finditer(r"archive/[^\s)]+\.md", archive_readme.read_text()):
            rel = m.group(0)
            if not (scope / rel).exists():
                findings.append({"kind": "archive_breadcrumb_dangling", "ref": rel})

    summary = {
        "scope": str(scope),
        "total_files_scanned": len(on_disk),
        "findings_count": len(findings),
        "by_kind": {},
    }
    for f in findings:
        summary["by_kind"][f["kind"]] = summary["by_kind"].get(f["kind"], 0) + 1

    if args.json:
        print(json.dumps({"summary": summary, "findings": findings}, indent=2))
        return 0

    print(f"nf-dream lint — {scope}")
    print(f"  files scanned: {summary['total_files_scanned']}")
    print(f"  findings:      {summary['findings_count']}")
    if not findings:
        print("  clean — no issues")
        return 0
    print()
    by_kind: dict[str, list[dict]] = {}
    for f in findings:
        by_kind.setdefault(f["kind"], []).append(f)
    for kind, items in by_kind.items():
        print(f"## {kind} ({len(items)})")
        for it in items[:30]:
            details = ", ".join(f"{k}={v}" for k, v in it.items() if k != "kind")
            print(f"  - {details}")
        if len(items) > 30:
            print(f"  ... and {len(items) - 30} more")
        print()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: nf-dream/scripts/lint.ts
#!/usr/bin/env bun
/**
 * nf-dream `lint` mode — read-only health check.
 *
 * Folder-wide regardless of project filter. Reports issues, never writes.
 *
 * Checks:
 *   - MEMORY.md size > 20 KB
 *   - Files > 30 KB
 *   - Missing frontmatter
 *   - Missing name/description/metadata.type/metadata.project/metadata.cwd
 *   - Misplaced top-level keys (type/project/cwd/tags/related at top instead of metadata:)
 *   - MEMORY.md links to non-existent files
 *   - Files on disk but missing from MEMORY.md
 *   - archive/README.md entries pointing to missing files
 *
 * Usage:
 *   bun lint.ts <scope> [--json]
 */
import { readFileSync, existsSync, statSync, readdirSync } from "node:fs";
import { resolve } from "node:path";
import { extractCwdSlugFromFilename, loadRegistry } from "./_lib";

const TOP_RESERVED = new Set([
  "name", "description", "created", "originSessionId", "last_read", "session_id",
]);
const METADATA_KEYS = new Set(["type", "project", "cwd", "tags", "related"]);
const SKIP_BASENAMES = new Set(["MEMORY.md"]);
const HANDOFF_PATTERN = /^HANDOFF([-_].+)?\.md$/i;

function parseFrontmatterLocal(text: string): { fmLines: string[] } | null {
  if (!text.startsWith("---\n")) return null;
  const lines = text.split("\n");
  let end: number | null = null;
  for (let i = 1; i < lines.length; i++) {
    if (lines[i] === "---") { end = i; break; }
  }
  if (end === null) return null;
  return { fmLines: lines.slice(1, end) };
}

interface FrontmatterAnalysis {
  has: Record<string, boolean>;
  misplacedTop: string[];
}

function analyzeFrontmatter(fmLines: string[]): FrontmatterAnalysis {
  const has = {
    name: false,
    description: false,
    metadata_block: false,
    "metadata.type": false,
    "metadata.project": false,
    "metadata.cwd": false,
  } as Record<string, boolean>;
  const misplacedTop: string[] = [];
  let mdIndentActive = false;
  let i = 0;
  while (i < fmLines.length) {
    const line = fmLines[i];
    if (line.trim() === "") { i++; continue; }
    const mTop = line.match(/^([A-Za-z_][A-Za-z0-9_]*):(.*)$/);
    const mNested = line.match(/^  ([A-Za-z_][A-Za-z0-9_]*):(.*)$/);
    if (mTop) {
      const key = mTop[1];
      mdIndentActive = false;
      if (key === "name") has.name = true;
      else if (key === "description") has.description = true;
      else if (key === "metadata") {
        has.metadata_block = true;
        mdIndentActive = true;
      } else if (METADATA_KEYS.has(key)) {
        misplacedTop.push(key);
      }
      i++;
      continue;
    }
    if (mNested && mdIndentActive) {
      const nkey = mNested[1];
      if (nkey === "type") has["metadata.type"] = true;
      else if (nkey === "project") has["metadata.project"] = true;
      else if (nkey === "cwd") has["metadata.cwd"] = true;
      i++;
      continue;
    }
    if (line.startsWith("  ") && mdIndentActive) { i++; continue; }
    mdIndentActive = false;
    i++;
  }
  return { has, misplacedTop };
}

interface Finding {
  kind: string;
  [key: string]: unknown;
}

interface Args {
  scope: string;
  json: boolean;
  help: boolean;
}

function parseArgs(argv: string[]): Args {
  const a: Args = { scope: "", json: false, help: false };
  const rest: string[] = [];
  for (const t of argv) {
    if (t === "--help" || t === "-h") a.help = true;
    else if (t === "--json") a.json = true;
    else if (t.startsWith("--")) throw new Error(`unknown flag: ${t}`);
    else rest.push(t);
  }
  if (rest.length === 0 && !a.help) throw new Error("missing <scope>");
  a.scope = rest[0] ?? "";
  return a;
}

function printHelp(): void {
  console.log(`bun lint.ts <scope> [--json]

Read-only health check on memory folder. Folder-wide regardless of project filter.`);
}

async function main(): Promise<number> {
  let args: Args;
  try {
    args = parseArgs(process.argv.slice(2));
  } catch (err) {
    console.error(`ERROR: ${(err as Error).message}`);
    printHelp();
    return 1;
  }
  if (args.help) { printHelp(); return 0; }
  const scope = resolve(args.scope.replace(/^~/, process.env.HOME ?? ""));
  try {
    if (!statSync(scope).isDirectory()) throw new Error();
  } catch {
    console.error(`ERROR: ${scope} is not a directory`);
    return 1;
  }

  const findings: Finding[] = [];

  const memoryMd = resolve(scope, "MEMORY.md");
  if (existsSync(memoryMd)) {
    const sizeKb = statSync(memoryMd).size / 1024;
    if (sizeKb > 20) {
      findings.push({ kind: "memory_md_large", size_kb: Math.round(sizeKb * 10) / 10 });
    }
  }

  const files = readdirSync(scope, { withFileTypes: true })
    .filter((e) => e.isFile() && e.name.endsWith(".md") && !SKIP_BASENAMES.has(e.name))
    .map((e) => ({ name: e.name, path: resolve(scope, e.name) }))
    .sort((a, b) => a.name.localeCompare(b.name));
  const onDisk = new Set(files.map((f) => f.name));

  const indexed = new Set<string>();
  const brokenLinks: string[] = [];
  if (existsSync(memoryMd)) {
    const memoryText = readFileSync(memoryMd, "utf8");
    for (const m of memoryText.matchAll(/\[[^\]]+\]\(([^)]+\.md)\)/g)) {
      const link = m[1];
      indexed.add(link);
      if (!existsSync(resolve(scope, link)) && !link.startsWith("archive/")) {
        brokenLinks.push(link);
      }
    }
  }

  const candidatesForIndex = new Set<string>();
  for (const n of onDisk) if (!HANDOFF_PATTERN.test(n)) candidatesForIndex.add(n);
  const missingIndex: string[] = [];
  for (const n of candidatesForIndex) if (!indexed.has(n)) missingIndex.push(n);
  missingIndex.sort();

  if (brokenLinks.length > 0) findings.push({ kind: "memory_md_broken_links", links: brokenLinks });
  if (missingIndex.length > 0) findings.push({ kind: "files_not_in_memory_md", files: missingIndex });

  const registry = loadRegistry();
  const knownSlugs = new Set([...Object.keys(registry.slugs), "generic"]);

  for (const { name, path } of files) {
    const sizeKb = statSync(path).size / 1024;
    if (sizeKb > 30) findings.push({ kind: "file_large", file: name, size_kb: Math.round(sizeKb * 10) / 10 });
    const text = readFileSync(path, "utf8");
    const parsed = parseFrontmatterLocal(text);
    if (!parsed) {
      findings.push({ kind: "no_frontmatter", file: name });
      continue;
    }
    const info = analyzeFrontmatter(parsed.fmLines);
    const missing = Object.entries(info.has).filter(([, v]) => !v).map(([k]) => k);
    if (missing.length > 0) findings.push({ kind: "missing_fields", file: name, missing });
    if (info.misplacedTop.length > 0) findings.push({ kind: "misplaced_top_level", file: name, keys: info.misplacedTop });

    // Filename suffix vs frontmatter.project consistency check.
    // Only enforce when BOTH are present; legacy files (no suffix) are not flagged.
    const candidate = extractCwdSlugFromFilename(name);
    const suffixSlug = candidate && (candidate === "generic" || knownSlugs.has(candidate))
      ? candidate
      : null;
    // Extract frontmatter project (single value, ignore array case for collision lint)
    let fmProject = "";
    for (const line of parsed.fmLines) {
      const m = line.match(/^  project:\s*(.*)$/);
      if (m) { fmProject = m[1].trim().replace(/^['"]|['"]$/g, ""); break; }
    }
    if (suffixSlug && fmProject && !fmProject.startsWith("[") && fmProject !== suffixSlug) {
      findings.push({
        kind: "filename_suffix_mismatch",
        file: name,
        frontmatter_project: fmProject,
        filename_suffix: suffixSlug,
      });
    }
  }

  const archiveReadme = resolve(scope, "archive", "README.md");
  if (existsSync(archiveReadme)) {
    for (const m of readFileSync(archiveReadme, "utf8").matchAll(/archive\/[^\s)]+\.md/g)) {
      const rel = m[0];
      if (!existsSync(resolve(scope, rel))) {
        findings.push({ kind: "archive_breadcrumb_dangling", ref: rel });
      }
    }
  }

  const byKind: Record<string, number> = {};
  for (const f of findings) byKind[f.kind] = (byKind[f.kind] ?? 0) + 1;

  const summary = {
    scope,
    total_files_scanned: onDisk.size,
    findings_count: findings.length,
    by_kind: byKind,
  };

  if (args.json) {
    console.log(JSON.stringify({ summary, findings }, null, 2));
    return 0;
  }

  console.log(`nf-dream lint — ${scope}`);
  console.log(`  files scanned: ${summary.total_files_scanned}`);
  console.log(`  findings:      ${summary.findings_count}`);
  if (findings.length === 0) { console.log("  clean — no issues"); return 0; }
  console.log();
  const grouped: Record<string, Finding[]> = {};
  for (const f of findings) {
    grouped[f.kind] = grouped[f.kind] ?? [];
    grouped[f.kind].push(f);
  }
  for (const [kind, items] of Object.entries(grouped)) {
    console.log(`## ${kind} (${items.length})`);
    for (const it of items.slice(0, 30)) {
      const details = Object.entries(it).filter(([k]) => k !== "kind").map(([k, v]) => `${k}=${JSON.stringify(v)}`).join(", ");
      console.log(`  - ${details}`);
    }
    if (items.length > 30) console.log(`  ... and ${items.length - 30} more`);
    console.log();
  }
  return 0;
}

if (import.meta.main) {
  main().then((code) => process.exit(code));
}
